src/counter/counter.go

```diff
@@ -0,0 +1,156 @@
+// vi:set tabstop=8 shiftwidth=8 noexpandtab:
+package counter
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/op/go-logging"
+
+	"github.com/nocive/go-request-counter/src/storage"
+	"github.com/nocive/go-request-counter/src/request"
+)
+
+type RequestCounter struct {
+	config  Config
+	bucket  request.RequestBucket
+	storage storage.RequestCounterStorage
+	logger  logging.Logger
+}
+
+func NewRequestCounter(c Config, b request.RequestBucket, s storage.RequestCounterStorage, l logging.Logger) *RequestCounter {
+	return &RequestCounter{
+		config:  c,
+		bucket:  b,
+		storage: s,
+		logger:  l,
+	}
+}
+
+func (r *RequestCounter) Init() error {
+	var err error
+
+	r.logger.Infof("booting!")
+
+	if !r.storage.Exists() {
+		r.logger.Info("data file doesn't exist, creating")
+		if err = r.storage.Create(); err != nil {
+			return err
+		}
+	} else {
+		r.logger.Info("data file exists, loading")
+		if err = r.storage.Load(&r.bucket); err != nil {
+			return err
+		}
+
+		c := r.bucket.GetCount()
+		r.logger.Infof("%d requests loaded from data file", c)
+
+		if (c > 0) {
+			r.logger.Info("firing a manual refresh")
+			c := r.bucket.Refresh() // purge expired events before starting
+			r.logger.Infof("purged %d requests from bucket", c)
+		}
+	}
+
+	r.logger.Infof(
+		"[ttl %s] [refresh %s] [save %s] [sleep %s] [count %d]",
+		r.config.RequestTtl,
+		r.config.RefreshInterval,
+		r.config.SaveInterval,
+		r.config.SleepPerRequest,
+		r.bucket.GetCount(),
+	)
+
+	r.logger.Info("initializing tickers and signal traps")
+	r.traps()
+
+	return nil
+}
+
+func (r *RequestCounter) Start() {
+	sema := make(chan struct{}, r.config.MaxClients)
+
+	http.HandleFunc("/", func(response http.ResponseWriter, request *http.Request) {
+		sema <- struct{}{}
+		defer func() { <-sema }()
+		r.process(response, request)
+	})
+
+	r.logger.Infof("preparing to listen on %s", r.config.BindAddr)
+	http.ListenAndServe(r.config.BindAddr, nil)
+}
+
+func (r *RequestCounter) process(response http.ResponseWriter, request *http.Request) {
```

Comment:
If we start to go by the book now, this should implement the `http.Handler` interface
https://golang.org/pkg/net/http/#Handler

Then the semaphore should become a reusable http middleware.
http://www.alexedwards.net/blog/making-and-using-middleware

Then create `http.Server` in the main function and `.ListenAndServe()` from there.